llama_index/vector_stores/bagel.py

```diff
@@ -0,0 +1,189 @@
+import logging
+import math
+from typing import Any, List
+
+from llama_index.schema import MetadataMode, TextNode
+from llama_index.vector_stores.types import (
+    MetadataFilters,
+    NodeWithEmbedding,
+    VectorStore,
+    VectorStoreQuery,
+    VectorStoreQueryResult,
+)
+from llama_index.vector_stores.utils import (
+    legacy_metadata_dict_to_node,
+    metadata_dict_to_node,
+    node_to_metadata_dict,
+)
+
+
+logger = logging.getLogger(__name__)
+
+
+def _to_bagel_filter(standard_filters: MetadataFilters) -> dict:
+    """
+    Translate standard metadata filters to Bagel specific spec.
+    """
+    filters = {}
+    for filter in standard_filters.filters:
+        filters[filter.key] = filter.value
+    return filters
+
+
+class BagelVectorStore(VectorStore):
+    """
+    Vector store for Bagel.
+    """
+
+    # support for Bagel specific parameters
+    stores_text: bool = True
+    flat_metadata: bool = True
+
+    def __init__(self, collection: Any, **kwargs: Any) -> None:
+        """
+        Initialize BagelVectorStore.
+
+        Args:
+            collection: Bagel collection.
+            **kwargs: Additional arguments.
+        """
+
+        try:
+            from bagel.api.Cluster import Cluster
+        except ImportError:
+            raise ImportError("Bagel is not installed. Please install bagel.")
+
+        if not isinstance(collection, Cluster):
+            raise ValueError("Collection must be a bagel Cluster.")
+
+        self._collection = collection
+
+    def add(
+        self, embedding_results: List[NodeWithEmbedding], **kwargs: Any
+    ) -> List[str]:
+        """
+        Add a list of nodes with embeddings to the vector store.
+
+        Args:
+            embedding_results: List of nodes with embeddings.
+            kwargs: Additional arguments.
+
+        Returns:
+            List of document ids.
+        """
+        if not self._collection:
+            raise ValueError("collection not set")
+
+        ids = []
+        embeddings = []
+        metadatas = []
+        documents = []
+
+        for node_with_embedding in embedding_results:
+            ids.append(node_with_embedding.id)
+            embeddings.append(node_with_embedding.embedding)
+            metadatas.append(
+                node_to_metadata_dict(
+                    node_with_embedding.node,
+                    remove_text=True,
+                    flat_metadata=self.flat_metadata,
+                )
+            )
+            documents.append(
+                node_with_embedding.node.get_content(metadata_mode=MetadataMode.NONE)
+                or ""
+            )
+
+        self._collection.add(
+            ids=ids, embeddings=embeddings, metadatas=metadatas, documents=documents
+        )
+
+        return ids
+
+    def delete(self, ref_doc_id: str, **kwargs: Any) -> None:
```

Comment:
Maybe a quick clarification, this is intended to delete according to a metadata field (`doc_id`), not the id of the node.

Is that possible with bagel?

Reply:
Yes. The document id retrieved from the bagel that is stored as a metadata. Not the node id.

Reply:
Hi, 
This is the node id. In bagel we create a cluster or collection of all nodes. We can delete any specific node using its id. 

My mistake. I will change it to node_id.

Reply:
Right, I guess all the other vector stores delete by `ref_doc_id`/`doc_id` rather than `node_id`

When a document is inserted into llama-index, the document object is broken into nodes. Each node has a new id, but the original document id is stored under `node.ref_doc_id`. Then when inserting nodes into vector stores, `node_to_metadata_dict` adds that ref doc id to the metadata.

All other vector stores operate by deleting using this original document id, since it is a more user-facing ID. If its not possible to delete by an ID that's in metadata, that's fine, but just wanted to confirm what's possible here.

Reply:
Thanks for the detailed explanation.. Is the current change okay?? 

BTW can you tell, why lint check is failing??

Reply:
Sorry, I think this is still incorrect. The id inserted into bagel should still be `result.id`

I took a look at the bagelDB docs. The delete function should probably look something like this? We find all IDs with the matching `doc_id`, then delete them

```
results = self._collection.get(where={"doc_id": ref_doc_id})
if results and 'id' in results:
  self._collection.delete(ids=results['id'])
```

Reply:
Please test this in the notebook though, just a guess after looking at the docs 😅

Reply:
@logan-markewich 

Thanks. Just ids, instead of id. Applied your suggested changes. 

now saving using id and deleting using ref_doc_id

Reply:
@logan-markewich 

Let me know if everything is okay now.. I will make the changes. Otherwise if you could merge the PR it will be appreciated. 
Thanks.